### PR TITLE
Lazy Template Loading - Nukie/Wiz

### LINKED ISSUE #17: Lazy load wizard's den, nuke ops base
Wizard's den and nuke ops base should only exist when they are actually needed.

Cost is currently unknown, but removing CTF maps saved several seconds.

Potential issue here is people meta'ing that these have spawned based on roundstart times, which did happen in the past with Clock Cult. I don't think it's an enormous deal as long as they keep it to themselves?

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Removes the nukie base and wizard den from the base centcom map. They are instead now lazy loaded as required.
To make a new lazy load template is fairly simple, make a map, allocate an area for it, and place a marker at the bottom left corner of that area. I have it check an area to ensure that if someone makes the map larger than expected but doesn't account for the template allocation it doesn't overwrite stuff without warning

[Replaces some improper CHECK_TICKs with MAPLOADING_CHECK_TICKs](https://github.com/tgstation/tgstation/pull/71785/commits/c7fbca9148812b392dca76d5450fb6bb86716462)

Atom init has already been tripped by the time we get to this portion of the loading, so if we don't use the right check tick, we will potentially block unrelated init attempts. This is bad. (Lemon edit, I want this in the commit desc)

## Why It's Good For The Game

Cuts down on init times.
Closes https://github.com/tgstation/dev-cycles-initiative/issues/17
## Changelog
:cl:
admin: New mapping verb to load lazy templates as needed. In your admin tab under the Mapping category.
/:cl:
this isnt technically player visible, so not sure it needs a changelog

### CODE REVIEW

File: code/controllers/subsystem/mapping.dm
Comment (san7890) [inline code/controllers/subsystem/mapping.dm:341]:
```suggestion
	loaded_lazy_templates = SSmapping.lazy_load_templates
```

yo code ain't compiling

**Comment by Mothblocks** [comment]
Need to measure how much this actually saves, and how noticable the init for players is when nukies roll

**Comment by Mothblocks** [comment]
Haven't checked yet but make sure that it loads when hijacked, but also that an admin can explicitly load it. Admins including myself have used the nuke ops base to prepare for a hijack we know is coming.

**Comment by ZephyrTFA** [comment]
Few more things to touch up

**Comment by JohnFulpWillard** [comment]
can we do this for abductors too (it doesn't really change much for init times but I think it would be nice to be consistent)

**Comment by Mothblocks** [comment]
I'll profile this when you have it passing CI

File: code/_globalvars/lists/mapping.dm
Comment (LemonInTheDark) [inline code/_globalvars/lists/mapping.dm:121]: this could be key -> pivot

File: code/controllers/subsystem/mapping.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/mapping.dm:862]: would yeet all of this

File: code/game/objects/effects/landmarks.dm
Comment (LemonInTheDark) [inline code/game/objects/effects/landmarks.dm:490]: why move the turfs to space?

File: code/game/objects/effects/landmarks.dm
Comment (LemonInTheDark) [inline code/game/objects/effects/landmarks.dm:481]: could we acomplish this with a variable on the landmark? loading non default space turfs is slower then just default ones

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/mapping.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/mapping.dm:862]: that or this should be a "load each of these pivots" pattern, rather then loading just one

File: code/controllers/subsystem/mapping.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/mapping.dm:851]: how appropriate is sleeping for like 10 seconds in the different places this is used?
loading post init will always take longer then loading pre init, because of changeturf issues

Comment (ZephyrTFA) [reply code/controllers/subsystem/mapping.dm:851]: I've noticed it's not actually taking very long to load the map. It's almost instant, locally atleast

File: code/game/objects/effects/landmarks.dm
Comment (ZephyrTFA) [reply code/game/objects/effects/landmarks.dm:490]: Didn't want the area to persist

Comment (ZephyrTFA) [reply code/game/objects/effects/landmarks.dm:490]: Which I now realize is stupid

File: code/game/objects/effects/landmarks.dm
Comment (ZephyrTFA) [reply code/game/objects/effects/landmarks.dm:481]: not without losing a sanity guard check against mappers increasing the size of a template without updating the pivot

**Comment by ZephyrTFA** [comment]
now also loads the nukie base if you select equip someone into nukie

File: code/game/objects/effects/landmarks.dm
Comment (LemonInTheDark) [reply code/game/objects/effects/landmarks.dm:481]: compare the size of the map to the size defined on the thing. a feature of the maploader is being able to dryload map files to get an idea of their size before placing them in the world

File: code/controllers/subsystem/mapping.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/mapping.dm:851]: interesting

File: code/_globalvars/lists/mapping.dm
Comment (LemonInTheDark) [inline code/_globalvars/lists/mapping.dm:121]: should doc the format

**Comment by ZephyrTFA** [comment]
alright reviews addressed and stuff works

File: code/modules/admin/verbs/mapping.dm
Comment (LemonInTheDark) [inline code/modules/admin/verbs/mapping.dm:494]: would be nice to have human readable names for these. actually  that could just BE the key but anyhow

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/mapping.dm
Comment (ZephyrTFA) [reply code/modules/admin/verbs/mapping.dm:498]:
```suggestion
	if(!length(GLOB.lazy_template_pivots[choice]))
```

File: code/modules/admin/verbs/mapping.dm
Comment (ZephyrTFA) [reply code/modules/admin/verbs/mapping.dm:494]: I'm personally using the format LT_NAME_OF_TEMPLATE

Comment (LemonInTheDark) [reply code/modules/admin/verbs/mapping.dm:494]: yes but it'd be nice to give admins like, proper names

**Comment by Mothblocks** [comment]
This saves about 0.4-0.7s. I was hoping for a lot more...but total init time for me is now about 33.4s when I merge master on top of this, so I'm happy as long as this is maintainable and a good framework. Still haven't looked at code tho.

**Comment by LemonInTheDark** [comment]
Can we do this to the ninja room too?

**Comment by Mothblocks** [comment]
also can you please explain what makes this system different from map_template?

**Comment by ZephyrTFA** [comment]
> also can you please explain what makes this system different from map_template?

what do you mean by this, it uses map templates to load stuff?

**Comment by ZephyrTFA** [comment]
> i would like to know the answers to the "can we" questions for the ninja room and abductors pod, but i dont think its necessary to put in this pr if its not trivial

its easy to do, just minorly time consuming; give me like an hour

**Comment by Mothblocks** [comment]
I guess what I'm asking is why is this a framework intruding in on SSmapping that also adds its own landmark and not *just* a map template? The SSmapping thing could just be to have a centralized place to pull these, and that is fine, but the landmark stuff weirds me out

**Comment by ZephyrTFA** [comment]
> I guess what I'm asking is why is this a framework intruding in on SSmapping that also adds its own landmark and not _just_ a map template? The SSmapping thing could just be to have a centralized place to pull these, and that is fine, but the landmark stuff weirds me out

There needs to be a landmark to actually signal where the lazy templates are going to be loaded at on the map itself. Otherwise there would be no way to locate where the template should be loaded

**Comment by ZephyrTFA** [comment]
> Why do we need to load them anywhere in particular, though? We don't care that the nuke base or wizard den is on CentCom, they're only there for historical reasons. I guess stuff checks is_centcom_level now but it just feels ready off, like we're going to eventually change the nuke ops base and then suddenly it eats into the range of other stuff.

Do you want me to add an empty Z level dedicated to lazy templates? And it'll keep track of used space accordingly?

**Comment by ZephyrTFA** [comment]
Probably need a subsystem to manage it though, SSlazy_mapping

**Comment by ZephyrTFA** [comment]
> > Probably need a subsystem to manage it though, SSlazy_mapping
> 
> You should just use the reservations system, the one that divides up empty z levels into usable space for loading maps

mothblocks wants me to have them be lazy maps instead, each lazy map will get a dedicated z level

**Comment by LemonInTheDark** [comment]
could honestly still use reservations for that (Tho a whole z level seems a bit overkill, and would for sure increase load times)

**Comment by Mothblocks** [comment]
I don't think they need to be a separate z-level, I just think they don't need to be in specific points at CentCom

**Comment by ZephyrTFA** [comment]
> I don't think they need to be a separate z-level, I just think they don't need to be in specific points at CentCom

How exactly do you expect me to accomplish finding a place to load them dynamically without being incredibly inefficient and bad

**Comment by Fikou** [comment]
so do they no longer spawn on centcom? thats disappointing.

**Comment by Mothblocks** [comment]
Yeah but also if they did what I asked and I think they did you'll be able to force them to spawn as an admin

Going to TM this tonight

**Comment by Mothblocks** [comment]
It really doesn't work at all. I spawned nuke ops and it spawned the base, but of the four nukies, only two actually showed up, were in their normal job outfits (and according to one, they were on station then teleported, even though there should be no delay), and the other two were in this weird limbo where they were stuck on the Join Game screen but were told they were nukies and I had to make new bodies and give them keys

**Comment by ZephyrTFA** [comment]
> It really doesn't work at all. I set nuke ops for roundstart on 196048 and it spawned the base, but of the four nukies, only two actually showed up, were in their normal job outfits (and according to one, they were on station then teleported, even though there should be no delay), and the other two were in this weird limbo where they were stuck on the Join Game screen but were told they were nukies and I had to make new bodies and give them keys

I'll look into it, probably missed something somewhere.

**Comment by Mothblocks** [comment]
will check tmrw when i can poop another nuke round

**Comment by ZephyrTFA** [comment]
abductors don't spawn correctly, ruleset needs to lazy load before pulling spawn points

**Comment by Mothblocks** [comment]
round 196251, zephyr joined a test merge

abductor spawned i  believe it was the agent on arrivals but the scientist back at home base, roundstart nuke ops worked fine (next time we'll test with midrounds), wizard didn't spawn because it couldn't find a point, but force spawning worked fine

**Comment by Mothblocks** [comment]
Also put in your changelog the new admin verb for spawning the stuff, it can be useful for pre-emptive hijacks

**Comment by san7890** [comment]
Forgot about this PR and merged #72014. My bad.

**Comment by Mothblocks** [comment]
its fine the problem is tooling not merging

**Comment by Mothblocks** [comment]
its not spawning roundstart wizards

**Comment by Mothblocks** [comment]
Eeeeee, causing some weird ass issues where we think it's conflicting with the arrivals shuttle on 196431

image

image

image

File: code/__HELPERS/_lists.dm
Comment (Mothblocks) [inline code/__HELPERS/_lists.dm:1103]: Document

File: code/__HELPERS/lazy_templates.dm
Comment (Mothblocks) [inline code/__HELPERS/lazy_templates.dm:1]: Document

`gen` -> `generate`

File: code/__HELPERS/lazy_templates.dm
Comment (Mothblocks) [inline code/__HELPERS/lazy_templates.dm:2]: `return .` at the end

File: code/datums/lazy_template.dm
Comment (Mothblocks) [inline code/datums/lazy_template.dm:22]:
```suggestion
		stack_trace("Something is trying to delete [type]")
		return QDEL_HINT_LETMELIVE
```

File: code/datums/lazy_template.dm
Comment (Mothblocks) [inline code/datums/lazy_template.dm:37]: I wouldn't bother with the lazy list here and would just default to `list()`, makes the rest of this uglier

File: code/game/area/areas/misc.dm
Comment (Mothblocks) [inline code/game/area/areas/misc.dm:38]: Do we still need this? I assume we are just loading map templates

File: code/modules/mapping/map_template.dm
Comment (Mothblocks) [inline code/modules/mapping/map_template.dm:155]: What's this?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/mapping/map_template.dm
Comment (ZephyrTFA) [reply code/modules/mapping/map_template.dm:155]: fixing the calculations for whether or not attempting to load a map at a specificed coord will go out of bounds.
width is one based not zero based so you need to subtract one from the calculation

**Comment by Mothblocks** [comment]
Happened again with a TM to get #72064, we gotta figure this out

**Comment by LemonInTheDark** [comment]
For context:
the shuttle's area is not being initialized. similarly, the shuttle's turfs aren't either. atoms are fine.

I have no idea why this is happening. Second time it happen it was a nukie round. Not sureif the first had any lateloading

**Comment by Mothblocks** [comment]
196507 didn't seem to have any nukies though :(

**Comment by JohnFulpWillard** [comment]
Nukie shuttle not loading breaks Syndicate headsets for traitors because they don't have a radio server

File: code/__DEFINES/lazy_templates.dm
Comment (Mothblocks) [inline code/__DEFINES/lazy_templates.dm:6]: Nit:

```suggestion
#define LAZY_TEMPLATE_KEY_LIST_ALL(...) list( \
```

...or make it a global list. Creating a list is not free or constant so it shouldn't be implied that it is.

File: code/__DEFINES/lazy_templates.dm
Comment (Mothblocks) [inline code/__DEFINES/lazy_templates.dm:13]: `__DEFINES` should only be defines, I know a lot of stuff does this right now but it annoys the hell out of me and regularly gets in the way. This should go either inside a `_globals` file, or with the rest of lazy template code.

**Comment by Addust** [comment]
what about NTnet's Wirecarp app? it has already proven able to remotely detect what White Ship it is, and if it sees the nukie base network i'm pretty sure any competent RD/captain/greyshit with a stolen ID would report it over radio

**Comment by Mothblocks** [comment]
can you elaborate on that? i dont know anything about wirecarp

also @LemonInTheDark has a fix for the blocking bug on this that they need to pr to the branch

**Comment by ZephyrTFA** [comment]
added syndicate mainframe to CC, also makes it to where people cant gamer WireCarp to see if there are nukies.
Also, someone should probably make wirecarp respect off network devices

File: code/controllers/subsystem/atoms.dm
Comment (Mothblocks) [inline code/controllers/subsystem/atoms.dm]: gut these before test merge

Comment (Mothblocks) [reply code/controllers/subsystem/atoms.dm]: oh i think you did

**Comment by san7890** [comment]
image

Not a fan of the verb placement, feels more for like the debug tab? Unsure.

**Comment by LemonInTheDark** [comment]
Fix is merged into this pr

**Comment by ZephyrTFA** [comment]
main issue work putting it in mapping verbs proper is that it will require you to have R_DEBUG, and i only care about R_ADMIN

**Comment by LemonInTheDark** [comment]
Honestly it's hard to event without vv, and so debug.
If people are so bad at hiring admins that they don't give them debug then I guess it sucks to be them

**Comment by Mothblocks** [comment]
@ZephyrTFA Can you move it to events or debug and then I can merge this

**Comment by ZephyrTFA** [comment]
Instructions unclear, new tab created explicitly for verb

**Pull request merged**
